lib/Sema/ConstraintPrinter.cpp

```diff
@@ -0,0 +1,381 @@
+        break;
+
+    case ConstraintKind::ExpressibleByBoolLiteral:
+        result = formatType(constraint->getSingleType())
+            + " : ExpressibleByBoolLiteral";
+        break;
+
+    case ConstraintKind::StructInitialiser:
+        result = "struct initializer for "
+            + formatType(constraint->getSingleType());
+        break;
+
+    case ConstraintKind::NumberOfConstraints:
+        result = "<invalid-constraint-kind>";
+        break;
+    }
+
+    // Add restriction info if available
+    if (constraint->hasRestriction()) {
+        result += " [restriction: ";
+        switch (constraint->getRestriction()) {
+        case ConversionRestrictionKind::DeepEquality:
+            result += "DeepEquality";
+            break;
+        case ConversionRestrictionKind::ArrayToPointer:
+            result += "ArrayToPointer";
+            break;
+        case ConversionRestrictionKind::StringToPointer:
+            result += "StringToPointer";
+            break;
+        case ConversionRestrictionKind::PointerToPointer:
+            result += "PointerToPointer";
+            break;
+        }
+        result += "]";
+    }
+
+    // Add flags
+    std::vector<std::string> flags;
+    if (constraint->isActive())
+        flags.push_back("active");
+    if (constraint->isDisabled())
+        flags.push_back("disabled");
+    if (constraint->isFavored())
+        flags.push_back("favored");
+    if (constraint->isDiscarded())
+        flags.push_back("discarded");
+    if (constraint->shouldRememberChoice())
+        flags.push_back("remember-choice");
+
+    if (!flags.empty()) {
+        result += " [";
+        for (size_t i = 0; i < flags.size(); ++i) {
+            if (i > 0)
+                result += ", ";
+            result += flags[i];
+        }
+        result += "]";
+    }
+
+    // Add source location info
+    if (constraint->getLocator()) {
+        result += " @<AST:0x";
+        result += llvm::Twine::utohexstr(
+                      reinterpret_cast<uintptr_t>(constraint->getLocator())
+        )
+                      .str();
+        result += ">";
+    }
+
+    return result;
+}
+
+std::string ConstraintPrinter::formatType(glu::types::Ty type)
+{
+    if (!type) {
+        return "<null>";
+    }
+
+    return _typePrinter.visit(type);
+}
+
+std::string
+ConstraintPrinter::formatFunctionDecl(glu::ast::FunctionDecl const *decl)
+{
+    if (!decl) {
+        return "<null-function>";
+    }
+
+    std::string result = decl->getName().str();
+
+    // Add function type information if available
+    if (auto *funcType
+        = llvm::dyn_cast<glu::types::FunctionTy>(decl->getType())) {
+        result += "(";
+        auto params = funcType->getParameters();
+        for (size_t i = 0; i < params.size(); ++i) {
+            if (i > 0)
+                result += ", ";
+            result += formatType(params[i]);
+        }
+        result += ") -> " + formatType(funcType->getReturnType());
+    }
```

Comment:
all of that is just formatType(decl->getType())... stop remaking what already exists